### ISSUE TITLE
Incorrect error message "use in later loop iteration"

### ISSUE THREAD

**Issue by mark-i-m** [rust-lang/rust#57959]
I have the following code (parsing the contents of `/proc/meminfo`):

```rust
                for line in reader {
                    // Split the line at the `:`
                    let split = line.expect("unable to read procfs").split(":");

                    let name = split.next().expect("Malformed input").trim(); // ERROR temporary in "split" is reused
                    let value = split.next().expect("Malformed input").trim();

                    map.insert(name, value);
                }
```

I correctly get an error that the temporary created for the first part of the expression in `split`'s RHS is reused on the following line.

However, the error message says that the borrow is used in a _later iteration of the loop_, which is false. The temporary is used later in the _same_ iteration.

```
error[E0716]: temporary value dropped while borrowed
   --> src/linux4_4/procfs/meminfo.rs:82:33
    |
82  |                       let split = line.expect("unable to read procfs").split(":");
    |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^           - temporary value is freed at the end of this statement
    |                                   |
    |                                   creates a temporary which is freed while still in use
83  |
84  |                       let name = split.next().expect("Malformed input").trim();
    |                                  ----- borrow used here, in later iteration of loop
```

**Comment by rep-nop** [comment]
Someone please correct me if I'm wrong (which is fairly likely lol), but this seems like a combination of two errors that are resulting in an incorrect error diagnostic. `BufRead::lines()` returns `std::io::Result<String>`, which then is being `split` generating an iterator over `&str`s which are then being put into what is presumably a `HashMap`, so the _correct_ error should be

```
error[E0597]: `split` does not live long enough
  --> src/main.rs:18:25
   |
18 |         let mut split = split.split(":");
   |                         ^^^^^ borrowed value does not live long enough
...
22 |         map.insert(name, value);
   |         --- borrow later used here
23 |     }
   |     - `split` dropped here while still borrowed
```
so maybe its being incorrectly reported as being later used because of the `dropped while borrowed` of both errors (there's also an error with `split` being immutable also, but not sure if that's relevant)? If anyone has any mentoring instructions on where this could be located (or a general idea of what might be happening) I'd be happy to fiddle around with a PR myself! :+1:

**Comment by mark-i-m** [comment]
That sounds plausible, but I'm not 100% sure. If you wanted to do some digging, the borrow checker errors are reported from here IIRC: https://github.com/rust-lang/rust/tree/master/src/librustc_mir/borrow_check/nll/region_infer/error_reporting

I'm sorry I can't be more helpful ATM...

**Comment by rep-nop** [comment]
No worries, I'll take a look around and see what I can find, thanks! :)

**Comment by steveklabnik** [comment]
I tried to reproduce this error, but couldn't quite: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=72a93b55e138ac96c9fbe9cf58d6642f

**Comment by steveklabnik** [comment]
Reproduction: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=822884a62cd60b75b47d8211a9b3e971


```rust
use std::io::{self, BufRead};
use std::collections::HashMap;

fn main() {
    
    let cursor = io::Cursor::new(b"lorem\nipsum\r\ndolor");
    let reader = cursor.lines();
    let mut map = HashMap::new();
    
    for line in reader {
        let mut split = line.unwrap().split(":");

        let name = split.next().unwrap().trim();
        let value = split.next().unwrap().trim();
        
        map.insert(name, value);
    }
}
```

gives:

```text
error[E0716]: temporary value dropped while borrowed
  --> src/main.rs:11:25
   |
11 |         let mut split = line.unwrap().split(":");
   |                         ^^^^^^^^^^^^^           - temporary value is freed at the end of this statement
   |                         |
   |                         creates a temporary which is freed while still in use
12 | 
13 |         let name = split.next().unwrap().trim();
   |                    ----- borrow later used here
   |
   = note: consider using a `let` binding to create a longer lived value
```

This no longer mentions "later in the loop", so I believe this has been fixed! Let me know if you're still seeing this. Thanks!
